File: Objective-C/tvOS/TabBarViewController.m

/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Tab bar view controller that manages the Products, Purchases, and Settings view controllers. Requests product information about a list of product
 identifiers using StoreManager. Listens and handles purchase and restore notifications.
*/

@import StoreKit;
#import "Utilities.h"
#import "Products.h"
#import "Purchases.h"
#import "StoreManager.h"
#import "StoreObserver.h"
#import "AppConfiguration.h"
#import "TabBarViewController.h"

typedef NS_ENUM(NSInteger, TabBarViewControllerItems) {
	TabBarViewControllerProducts = 0,
	TabBarViewControllerPurchases,
	TabBarViewControllerSettings
};

@interface TabBarViewController () <UITabBarControllerDelegate>
@property (strong) Utilities *utility;
@property BOOL restoreWasCalled;

@property (strong) Products *products;
@property (strong) Purchases *purchases;
@end

@implementation TabBarViewController
#pragma mark - Initializers

-(instancetype)initWithCoder:(NSCoder *)coder {
	self = [super initWithCoder:coder];
	if (self) {
		_utility = [[Utilities alloc] init];
		_restoreWasCalled = NO;
	}
	return self;
}

#pragma mark - Instantiate View Controllers

-(Products *)instantiateProducts {
	UINavigationController *navigation = (UINavigationController *)self.viewControllers[TabBarViewControllerProducts];
	if ([navigation.topViewController isKindOfClass:[Products class]]) {
		Products *controller = (Products *)navigation.topViewController;
		return controller;
	}
	return nil;
}

-(Purchases *)instantiatePurchases {
	UINavigationController *navigation = (UINavigationController *)self.viewControllers[TabBarViewControllerPurchases];
	if ([navigation.topViewController isKindOfClass:[Purchases class]]) {
		Purchases *controller = (Purchases *)navigation.topViewController;
		return controller;
	}
	return nil;
}

#pragma mark - View Life Cycle

- (void)viewDidLoad {
	[super viewDidLoad];
	self.delegate = self;

	self.products = [self instantiateProducts];
	self.purchases = [self instantiatePurchases];

	// Registers for product request, purchase, and restore notifications.
	[[NSNotificationCenter defaultCenter] addObserver:self
											 selector:@selector(handleProductRequestNotification:)
												 name:PCSProductRequestNotification
											   object:[StoreManager sharedInstance]];

	[[NSNotificationCenter defaultCenter] addObserver:self
											 selector:@selector(handlePurchaseNotification:)
												 name:PCSPurchaseNotification
											   object:[StoreObserver sharedInstance]];

	[[NSNotificationCenter defaultCenter] addObserver:self
											 selector:@selector(handleRestoreNotification:)
												 name:PCSRestoredWasCalledNotification
											   object:nil];
}

#pragma mark - Fetch Product Information

/// Retrieves product information from the App Store.
-(void)fetchProductInformation {
	// First, let's check whether the user is allowed to make purchases. Proceed if they are allowed. Display an alert, otherwise.
	if ([StoreObserver sharedInstance].isAuthorizedForPayments) {
		NSArray *identifiers = self.utility.identifiers;

		if (identifiers != nil) {
			if (identifiers.count > 0) {
				// Refresh the UI with identifiers to be queried.
				Section *section = [[Section alloc] initWithName:PCSProductsInvalidIdentifiers elements:identifiers];
				[self.products reloadWithData:@[section]];

				// Fetch the product information.
				[[StoreManager sharedInstance] startProductRequestWithIdentifiers:identifiers];
			} else {
				// Warn the user that the resource file does not contain anything.
				[self alertWithTitle:PCSMessagesStatus message:[NSString stringWithFormat:@"%@.%@ %@", PCSProductIdsPlistName, PCSProductIdsPlistFileExtension, PCSMessagesEmptyResource]];
			}
		} else {
			// Warn the user that the resource file could not be found.
			[self alertWithTitle:PCSMessagesStatus message:[NSString stringWithFormat:@"%@ %@.%@.", PCSMessagesResourceNotFound, PCSProductIdsPlistName, PCSProductIdsPlistFileExtension]];
		}
	} else {
		// Warn the user that they are not allowed to make purchases.
		[self alertWithTitle:PCSMessagesStatus message:[NSString stringWithFormat:@"%@", PCSMessagesCannotMakePayments]];
	}
}

#pragma mark - Handle PCSProductRequest Notification

/// Updates the UI according to the product request notification result.
-(void)handleProductRequestNotification:(NSNotification *)notification {
	StoreManager *productRequestNotification = (StoreManager*)notification.object;
	PCSProductRequestStatus status = (PCSProductRequestStatus)productRequestNotification.status;

	if (status == PCSStoreResponse) {
		[self.products reloadWithData:productRequestNotification.storeResponse];
	} else if (status == PCSRequestFailed) {
		[self alertWithTitle:PCSMessagesProductRequestStatus message:productRequestNotification.message];
	}
}

#pragma mark - Handle PCSPurchase Notification

/// Updates the UI according to the purchase request notification result.
-(void)handlePurchaseNotification:(NSNotification *)notification {
	StoreObserver *purchasesNotification = (StoreObserver *)notification.object;
	PCSPurchaseStatus status = (PCSPurchaseStatus)purchasesNotification.status;

	switch (status) {
		case PCSNoRestorablePurchases:
		case PCSPurchaseFailed:
		case PCSRestoreFailed:
			[self alertWithTitle:PCSMessagesPurchaseStatus message:purchasesNotification.message];
			break;
		// Switch to the Purchases view when receiving a successful restore notification.
		case PCSRestoreSucceeded: [self handleRestoredSucceededTransaction];
			break;
		default: break;
	}
}

#pragma mark - Handle Restored Transactions

/// Handles succesful restored transactions. Switches to the Purchases tab.
-(void)handleRestoredSucceededTransaction {
	self.utility.restoreWasCalled = self.restoreWasCalled;
	[self.purchases reloadWithData:self.utility.dataSourceForPurchasesUI];
	self.selectedIndex = TabBarViewControllerPurchases;
}

#pragma mark - Handle PCSRestoredWasCalled Notification

-(void)handleRestoreNotification:(NSNotification *)notification {
	self.restoreWasCalled = YES;
}

#pragma mark - Managing Tab Bar Selections

-(BOOL)tabBarController:(UITabBarController *)tabBarController shouldSelectViewController:(UIViewController *)viewController {
	return [StoreObserver sharedInstance].isAuthorizedForPayments;
}

-(void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController {
	NSInteger index = [tabBarController.viewControllers indexOfObject:viewController];

	if (index == TabBarViewControllerProducts) {
		self.restoreWasCalled = NO;
		[self fetchProductInformation];
	} else if (index == TabBarViewControllerPurchases) {
		UINavigationController *navigation = (UINavigationController *)viewController;

		if ([navigation.topViewController isKindOfClass:[Purchases class]]) {
			Purchases *controller = (Purchases *)navigation.topViewController;
			self.utility.restoreWasCalled = self.restoreWasCalled;
			[controller reloadWithData:self.utility.dataSourceForPurchasesUI];
		}
	} else if (index == TabBarViewControllerSettings) {
		self.restoreWasCalled = NO;
	}
}

#pragma mark - Display Alert

/// Creates and displays an alert.
-(void)alertWithTitle:(NSString *)title message:(NSString *)message {
	UIAlertController *alertController = [self.utility alertWithTitle:title message:message];
	[self presentViewController:alertController animated:YES completion:nil];
}


#pragma mark - Memory Management

- (void)dealloc {
	// Unregister for StoreManager's notifications.
	[[NSNotificationCenter defaultCenter] removeObserver:self name:PCSProductRequestNotification object:[StoreManager sharedInstance]];

	// Unregister for StoreObserver's notifications.
	[[NSNotificationCenter defaultCenter] removeObserver:self name:PCSPurchaseNotification object:[StoreObserver sharedInstance]];

	// Unregister for Settings' notifications.
	[[NSNotificationCenter defaultCenter] removeObserver:self name:PCSRestoredWasCalledNotification object:nil];
}

@end

File: Swift/tvOS/TabBarViewController.swift
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Tab bar view controller that manages the Products, Purchases, and Settings view controllers. Requests product information about a list of product
identifiers using StoreManager. Listens and handles purchase and restore notifications.
*/

import UIKit
import StoreKit

class TabBarViewController: UITabBarController {
	// MARK: - Types

	enum TabBarViewControllerItems: Int {
		case products, purchases, settings
	}

	// MARK: - Properties

	fileprivate var utility = Utilities()
	fileprivate var resourceFile = ProductIdentifiers()
	fileprivate var restoreWasCalled = false

	fileprivate lazy var products: Products = {
		guard let navigation = self.viewControllers?[TabBarViewControllerItems.products.rawValue] as? UINavigationController
			else { fatalError("\(Messages.unableToInstantiateNavigationController)") }

		guard let controller = navigation.topViewController as? Products else { fatalError("\(Messages.unableToInstantiateProducts)") }
		return controller
	}()

	fileprivate lazy var purchases: Purchases = {
		guard let navigation = self.viewControllers?[TabBarViewControllerItems.purchases.rawValue] as? UINavigationController
			else { fatalError("\(Messages.unableToInstantiateNavigationController)") }

		guard let controller = navigation.topViewController as? Purchases else { fatalError("\(Messages.unableToInstantiatePurchases)") }
		return controller
	}()

	fileprivate lazy var settings: Settings = {
		guard let navigation = self.viewControllers?[TabBarViewControllerItems.settings.rawValue] as? UINavigationController
			else { fatalError("\(Messages.unableToInstantiateNavigationController)") }

		guard let controller = navigation.topViewController as? Settings else { fatalError("\(Messages.unableToInstantiateSettings)") }
		return controller
	}()

	// MARK: - View Life Cycle

	override func viewDidLoad() {
		super.viewDidLoad()
		self.delegate = self

		StoreManager.shared.delegate = self
		StoreObserver.shared.delegate = self
		settings.delegate = self
	}

	// MARK: - Fetch Product Information

	/// Retrieves product information from the App Store.
	fileprivate func fetchProductInformation() {
		// First, let's check whether the user is allowed to make purchases. Proceed if they are allowed. Display an alert, otherwise.
		if StoreObserver.shared.isAuthorizedForPayments {
			let resourceFile = ProductIdentifiers()

			guard let identifiers = resourceFile.identifiers else {
				// Warn the user that the resource file could not be found.
				alert(with: Messages.status, message: resourceFile.wasNotFound)
				return
			}

			if !identifiers.isEmpty {
				// Refresh the UI with identifiers to be queried.
				products.reload(with: [Section(type: .invalidProductIdentifiers, elements: identifiers)])

				// Fetch the product information.
				StoreManager.shared.startProductRequest(with: identifiers)
			} else {
				// Warn the user that the resource file does not contain anything.
				alert(with: Messages.status, message: resourceFile.isEmpty)
			}
		} else {
			// Warn the user that they are not allowed to make purchases.
			alert(with: Messages.status, message: Messages.cannotMakePayments)
		}
	}

	// MARK: - Handle Restored Transactions

	/// Handles succesful restored transactions. Switches to the Purchases tab.
	fileprivate func handleRestoredSucceededTransaction() {
		utility.restoreWasCalled = restoreWasCalled
		purchases.reload(with: utility.dataSourceForPurchasesUI)
		selectedIndex = 1
	}

	// MARK: - Display Alert

	/// Creates and displays an alert.
	func alert(with title: String, message: String) {
		let alertController = utility.alert(title, message: message)
		self.present(alertController, animated: true, completion: nil)
	}
}

// MARK: - UITabBarControllerDelegate

/// Extends TabBarViewController to conform to UITabBarControllerDelegate.
extension TabBarViewController: UITabBarControllerDelegate {
	func tabBarController(_ tabBarController: UITabBarController, shouldSelect viewController: UIViewController) -> Bool {
		return StoreObserver.shared.isAuthorizedForPayments
	}

	func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
		guard let index = tabBarController.viewControllers?.index(of: viewController) else { fatalError("\(Messages.unknownSelectedViewController)") }

		guard let item = TabBarViewControllerItems(rawValue: index) else { fatalError("\(Messages.unknownTabBarIndex)\(index)).") }

		switch item {
		case .products:
			restoreWasCalled = false
			fetchProductInformation()
		case .purchases:
			if let controller = (viewController as? UINavigationController)?.topViewController as? Purchases {
				utility.restoreWasCalled = restoreWasCalled
				controller.reload(with: utility.dataSourceForPurchasesUI)
			}
		case .settings: restoreWasCalled = false
		}
	}
}

// MARK: - SettingsDelegate

/// Extends TabBarViewController to conform to SettingsDelegate.
extension TabBarViewController: SettingsDelegate {
	func settingDidSelectRestore() {
		restoreWasCalled = true
	}
}

// MARK: - StoreManagerDelegate

/// Extends TabBarViewController to conform to StoreManagerDelegate.
extension TabBarViewController: StoreManagerDelegate {
	func storeManagerDidReceiveResponse(_ response: [Section]) {
		products.reload(with: response)
	}

	func storeManagerDidReceiveMessage(_ message: String) {
		alert(with: Messages.productRequestStatus, message: message)
	}
}

// MARK: - StoreObserverDelegate

/// Extends TabBarViewController to conform to StoreObserverDelegate.
extension TabBarViewController: StoreObserverDelegate {
	func storeObserverDidReceiveMessage(_ message: String) {
		alert(with: Messages.purchaseStatus, message: message)
	}

	func storeObserverRestoreDidSucceed() {
		handleRestoredSucceededTransaction()
	}
}

